### ISSUE TITLE
Empty `String[]`: object is not an instance of declaring class

### ISSUE THREAD

**Issue by cheako** [astonbitecode/j4rs#50]
```rust
        let node_starter = jvm.create_instance("freenet.node.NodeStarter", &[])?;
        dbg!(jvm
            .invoke(
                &node_starter,
                "start",
                &[jvm.create_java_array("java.lang.String", &[])?.into()]
            )?
            .class_name());
```

```java
	public Integer start(String[] args) {
```

Also, should I open another issue or can you not `impl Debug for Instance && InstanceArg`?  Obviously what I'd be looking for is json + comments denoting any `type` or instance information not included in json + a flag for `Arg`.

Edit: FWIW, I have the java source and can modify/recompiled it.

**Comment by cheako** [comment]
Sorry, forgot the error:
```text
   Compiling truffle v0.1.0 (/home/cheako/src/gitlab/truffle)
    Finished test [unoptimized + debuginfo] target(s) in 0.34s
     Running unittests src/main.rs (target/debug/deps/truffle-771c7bd7f5fdff9f)

running 1 test
Exception in thread "main" org.astonbitecode.j4rs.errors.InvocationException: While invoking method start of Class freenet.node.NodeStarter
        at org.astonbitecode.j4rs.api.invocation.JsonInvocationImpl.invoke(JsonInvocationImpl.java:73)
Caused by: java.lang.IllegalArgumentException: object is not an instance of declaring class
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.astonbitecode.j4rs.api.invocation.JsonInvocationImpl.invokeMethod(JsonInvocationImpl.java:184)
        at org.astonbitecode.j4rs.api.invocation.JsonInvocationImpl.invoke(JsonInvocationImpl.java:70)
Error: JavaError("An Exception was thrown by Java... Please check the logs or the console.")
```

**Comment by astonbitecode** [comment]
```rust
dbg!(jvm.invoke(
      &node_starter,
      "start",
      &[jvm.create_java_array("java.lang.String", &[])?.into()]
   )?.class_name());
```

gives the output

> jvm.invoke(&node_starter, "start",
>            &[jvm.create_java_array("java.lang.String",
>                                   &[])?.into()])?.class_name() = "known_in_java_world"

How did you get the error you mentioned?

The class name is not known in the Rust world, as the `Instance` returned from the `jvm.invoke(...)` is the outcome of a Java invocation. 

It would be possible to get the class name of the actual Java Object that the `Instance` holds, but this would require JNI operation, having performance impact. 

What are you trying to achieve? Don't you always know which is the class that should be returned from an invocation?

**Comment by cheako** [comment]
Ignore the class_name part, obviously from the prototype it's going to be an `Intager`.  What I wanted was for `dbg!`/`Debug` to print something like `Integer { 7 }`.  but `Instance` doesn't implement Debug, so I called a function at random that does return a type that would at least print something.  Though that's besides the point because I don't think `class_name` is even being called.

**Comment by cheako** [comment]
As suspected...
```rust

        let node_starter = jvm.create_instance("freenet.node.NodeStarter", &[])?;
        let int = jvm.invoke(
            &node_starter,
            "start",
            &[jvm.create_java_array("java.lang.String", &[])?.into()],
        )?;
        panic!("Did we make it?");
        dbg!(int.class_name());
```
Does not panic with "Did we make it?", instead I get the same error/above.

**Comment by astonbitecode** [comment]
The `class_name` is called in my case as you see in my output, but it does not bring any value because it prints

> known_in_java_world 

Your code fails with this error but I cannot say why. Can you provide some more details of your setup? OS, Java version, Rust version and anything more would help. If you have some code to share please do so too.

About debugging `Instance`s, as I said, it can be done, but maybe not by default (need to investigate the performance impact) because it implies additional JNI call. Maybe we can create a method like eg: `jvm.debug_instance(&Instance)` or something similar and users can explicitly call it, understanding the possible impact.

**Comment by cheako** [comment]
I'm unsure of how to get java version, I actually know next to nothing about java...  Just that the types of the arguments effect what line number is run next, so I need to make sure the argument is `String[]`.

```text
cheako@mx1:~$ rustc --version
rustc 1.69.0-nightly (e972bc808 2023-01-29)
cheako@mx1:~$ java --version
openjdk 17.0.6 2023-01-17
OpenJDK Runtime Environment (build 17.0.6+10-Debian-1deb11u1)
OpenJDK 64-Bit Server VM (build 17.0.6+10-Debian-1deb11u1, mixed mode, sharing)
cheako@mx1:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye
```

```rust
#[cfg(test)]
mod tests {
    #[allow(unused_imports)]
    use j4rs::prelude::*;
    #[allow(unused_imports)]
    use j4rs::InvocationArg;
    use j4rs::JvmBuilder;
    #[allow(unused_imports)]
    use j4rs_derive::*;

    #[test]
    #[allow(unreachable_code, unused_variables)]
    fn one() -> Result<(), Box<dyn std::error::Error>> {
        // std::env::set_current_dir("/home/srv_fn/var");

        let jvm = JvmBuilder::new().build()?;

        jvm.deploy_artifact(&j4rs::LocalJarArtifact::new("freenet.jar"))?;
        jvm.deploy_artifact(&j4rs::LocalJarArtifact::new("bcprov-jdk15on-1.59.jar"))?;
        jvm.deploy_artifact(&j4rs::LocalJarArtifact::new("wrapper.jar"))?;
        jvm.deploy_artifact(&j4rs::LocalJarArtifact::new("freenet-ext.jar"))?;

        let node_starter = jvm.create_instance("freenet.node.NodeStarter", &[])?;
        let int = jvm.invoke(
            &node_starter,
            "start",
            &[jvm.create_java_array("java.lang.String", &[])?.into()],
        )?;
        panic!("Did we make it?");
        dbg!(int.class_name());

        // let r0 = jvm.create_instance("freenet.crypt.Yarrow", &[])?;
        /* let r1 = jvm.create_instance(
            "freenet.crypt.Yarrow",
            &[
                "prng.seed".try_into()?,
                "SHA1".try_into()?,
                "Rijndael".try_into()?,
                InvocationArg::into_primitive(false.try_into()?)?,
                InvocationArg::into_primitive(false.try_into()?)?,
            ],
        )?;
        */
        // let node = jvm.create_instance("freenet.Node", &[todo!(), r0.into(), r1.into()])?;
        // let fpm = jvm.create_instance("freenet.node.FNPPacketMangler", &[])?;

        Ok(())
    }
}
```
[jar_files.zip](https://github.com/astonbitecode/j4rs/files/10807141/jar_files.zip)

**Comment by astonbitecode** [comment]
I cannot know how `freenet.node.NodeStarter` actually works, but as I see [here](https://github.com/freenet/fred/blob/ffe2268f5d9c42aae58476c1559cdeb6a4ca8e2d/src/freenet/node/NodeStarter.java#L77) the constructor that you call with 
```rust
let node_starter = jvm.create_instance("freenet.node.NodeStarter", &[])?;
```
is __private__.

This means that you don't actually instantiate a `NodeStarter` object and this is why you get this error when you try to invoke `NodeStarter#start`.

Thus, this does not seem a j4rs issue.

About the debug info, I honestly don't find it necessary at this point, at least not until we have a concrete case where it can be useful.

**Comment by cheako** [comment]
Thanks.
